vim-utils: fix dependency graph

baloo opened the pull request:
###### Motivation for this change

when #138349 got merged, it broke the dependency graph between the vim
config and its plugins.

By rewriting the path it seems the edge gets removed. This makes the
configuration file inherit a correct path but without the plugin backing
it.

See https://github.com/NixOS/nixpkgs/pull/138349#issuecomment-931907960
cc @anirudhb @teto

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

figsoda commented:
maybe we can add some tests for dependency resolution

@ofborg build neovim.tests

baloo commented:
@figsoda not too sure how to implement such a test. Any pointers.

baloo committed:
neovim.tests: ensure the graph is pulled

We'll change the plugin name to ensure the plugin is correctly pulled by
vim-plug config generation.

Signed-off-by: Arthur Gautier <baloo@superbaloo.net>

baloo commented:
actually the `nix-build . -A neovim.tests.run_nvim_with_plug` test was already covering this. I made a small change to change the name and ensure we don't run in false positives there.

teto commented:
hum the test with base16-vim used to succeed, why is that ? 

The initial problem was that 
```
Plug '/nix/store/xxx-vimplugin/./` 
```
failed to work with vim plug because of the trailing "./". Wouldn't just not writing the last part be enough instead of this complex stuff ?instead of having `stripDots pkg.rtp` just use `pkg` in plugImpl.
Alternatively one could try to fix vim-plug too.

baloo committed:
neovim.tests: ensure nvim exit with error

Signed-off-by: Arthur Gautier <baloo@superbaloo.net>

baloo commented:
Current master @ 810ed7a053191b5f8c9700dcf8b21eef5a28e887 is broken:
```
$ nix-build . -A neovim.tests.run_nvim_with_plug
this derivation will be built:
  /nix/store/2sqmbn45hiwjqinsdk6gvqi6g3jsyr14-test-neovim-0.5.1-with-plug.drv
building '/nix/store/2sqmbn45hiwjqinsdk6gvqi6g3jsyr14-test-neovim-0.5.1-with-plug.drv'...
=11;?(B112112Error detected while processing /nix/store/chxzgcnfb704x4jpafjrnazn7qsg9kpa-ini                                                 

t.vim:
line   10:
E185: Cannot find color scheme 'base16-tomorrow-night'112



Error detected while processing command line:

112(B>
/nix/store/k1ci4rxb8rqqavf88r2al59vgbhnrv33-test-neovim-0.5.1-with-plug

$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

and it looks like the tests didn't check anything:
```
$ nix-build . -A neovim.tests.run_nvim_with_plug
this derivation will be built:
  /nix/store/2sqmbn45hiwjqinsdk6gvqi6g3jsyr14-test-neovim-0.5.1-with-plug.drv
building '/nix/store/2sqmbn45hiwjqinsdk6gvqi6g3jsyr14-test-neovim-0.5.1-with-plug.drv'...
=11;?(B112112Error detected while processing /nix/store/chxzgcnfb704x4jpafjrnazn7qsg9kpa-ini                                                 

t.vim:
line   10:
E185: Cannot find color scheme 'base16-tomorrow-night'112



Error detected while processing command line:

112(B>
/nix/store/k1ci4rxb8rqqavf88r2al59vgbhnrv33-test-neovim-0.5.1-with-plug

$ echo $?
0
```

I've pushed a new version which fixes the test.

I have a limited knowledge of vim scripts, not sure I can fix vim-plug itself.

SuperSandro2000 commented on pkgs/applications/editors/neovim/tests.nix:
```diff
@@ -87,7 +87,7 @@ rec {
     extraName = "-with-plug";
     configure.plug.plugins = with pkgs.vimPlugins; [
-      base16-vim
+      (base16-vim.overrideAttrs(old: { pname=old.pname + "-unique-for-tests-please-dont-use"; }))
```
```suggestion
      (base16-vim.overrideAttrs(old: { pname = old.pname + "-unique-for-tests-please-dont-use"; }))
```

baloo committed:
Revert "fix: remove trailing '/.' from vim-plug plugin paths"

The root cause if fixed by
https://github.com/nixOS/nixpkgs/commit/761b2c6ff342425388db38b86f1c49cbdb232514#diff-e4b94db9201d58bd9410739dddf92bef74e0b5f5e596c804a84ee7c580ae3f71R9

This reverts commit d9d1a11aed1761bc3a9d95dd2b8d154c44d4feec.

baloo commented:
@teto the root cause of the "/./." in the path ended up being fixed by https://github.com/nixOS/nixpkgs/commit/761b2c6ff342425388db38b86f1c49cbdb232514#diff-e4b94db9201d58bd9410739dddf92bef74e0b5f5e596c804a84ee7c580ae3f71R9

I just reverted the fix introduced in #138349 and fixed the tests to ensure we do not introduce the issue again.

teto commented on pkgs/applications/editors/neovim/tests.nix:
```diff
@@ -87,7 +87,7 @@ rec {
     extraName = "-with-plug";
     configure.plug.plugins = with pkgs.vimPlugins; [
-      base16-vim
+      (base16-vim.overrideAttrs(old: { pname = old.pname + "-unique-for-tests-please-dont-use"; }))
```
can we get rid of that ?

baloo commented:
@teto bump.

samueldr commented:
I hadn't seen this PR while trying to figure out why no plugins were loading, here's an alternative implementation that keeps the `split` method, in case anyone has a similar issue and wants an alternative solution

```diff
From c16917bc6d5577bf51a5ed1e90a62e47e561f56d Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sun, 10 Oct 2021 16:05:59 -0400
Subject: [PATCH] vimPlugins: fix Plug losing string context

This would, in turn, mean store path references wouldn't stick when
copying closures.

This was observed in a setup where a build machine is used with
`--build-host` to build the system configuration.

Reverting d9d1a11aed1761bc3a9d95dd2b8d154c44d4feec brought back the
references.
---
 pkgs/misc/vim-plugins/vim-utils.nix | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pkgs/misc/vim-plugins/vim-utils.nix b/pkgs/misc/vim-plugins/vim-utils.nix
index 60b9b60806d..0672455ddcc 100644
--- a/pkgs/misc/vim-plugins/vim-utils.nix
+++ b/pkgs/misc/vim-plugins/vim-utils.nix
@@ -293,7 +293,7 @@ let
       /* vim-plug is an extremely popular vim plugin manager.
       */
       /* Remove repeated "/." suffixes from a path */
-      stripDots = path: lib.head (builtins.split "(/\\.)*$" path);
+      stripDots = path: lib.addContextFrom path (lib.head (builtins.split "(/\\.)*$" path));
       plugImpl =
       (''
         source ${vimPlugins.vim-plug.rtp}/plug.vim
-- 
2.33.0
```

baloo commented:
I didn't know about addContextFrom, this is great!
